src/uu/cp/src/cp.rs

```diff
@@ -627,46 +651,81 @@ impl CopyMode {
     }
 }
 
-impl FromStr for Attribute {
-    type Err = Error;
+impl Attributes {
+    // TODO: ownership is required if the user is root, for non-root users it's not required.
+    // See: https://github.com/coreutils/coreutils/blob/master/src/copy.c#L3181
 
-    fn from_str(value: &str) -> CopyResult<Self> {
-        Ok(match &*value.to_lowercase() {
-            "mode" => Self::Mode,
+    fn all() -> Self {
+        Self {
             #[cfg(unix)]
-            "ownership" => Self::Ownership,
-            "timestamps" => Self::Timestamps,
-            #[cfg(feature = "feat_selinux")]
-            "context" => Self::Context,
-            "links" => Self::Links,
-            "xattr" => Self::Xattr,
+            ownership: Preserve::Yes { required: true },
+            mode: Preserve::Yes { required: true },
+            timestamps: Preserve::Yes { required: true },
+            context: {
+                #[cfg(feature = "feat_selinux")]
+                {
+                    Preserve::Yes { required: false }
+                }
+                #[cfg(not(feature = "feat_selinux"))]
+                {
+                    Preserve::No
+                }
+            },
+            links: Preserve::Yes { required: true },
+            xattr: Preserve::Yes { required: false },
+        }
+    }
+
+    fn default() -> Self {
+        Self {
+            #[cfg(unix)]
+            ownership: Preserve::Yes { required: true },
+            mode: Preserve::Yes { required: true },
+            timestamps: Preserve::Yes { required: true },
+            context: Preserve::No,
+            links: Preserve::No,
+            xattr: Preserve::No,
+        }
+    }
+
+    fn none() -> Self {
+        Self {
+            #[cfg(unix)]
+            ownership: Preserve::No,
+            mode: Preserve::No,
+            timestamps: Preserve::No,
+            context: Preserve::No,
+            links: Preserve::No,
+            xattr: Preserve::No,
+        }
+    }
+
+    /// Tries to match string containing a parameter to preserve with the corresponding entry in the
+    /// Attributes struct. The values are maximized, which means that, when the Attributes are
+    /// initialized with some preserve levels, and later found in the `--preserve=...` values, they
+    /// are updated to maximum of the one set already and the one found in the string from the user.
+    fn try_set_from_string(&mut self, value: &str) -> Result<(), Error> {
+        let preserve_yes_required = Preserve::Yes { required: true };
+
+        match &*value.to_lowercase() {
+            "mode" => self.mode = self.mode.max(preserve_yes_required),
```

Comment:
We don't need `max` here, right? `self.mode = preserve_yes_required` should be enough.

Reply:
You are right, updated!